accepted/2020/serializer/WriteableDomAndDynamic.md

```diff
@@ -0,0 +1,587 @@
+```
+
+### Cast operators
+No explicit or implicit cast operators exist since `GetValue<T>` allows for a single, consistent programming model. This also works on not just the "known" types in the CLR but also custom data types known by the serializer.
+
+Cast operators do not work in all languages such as F# and even for C# would not work for all types:
+```cs
+JsonArray jArray = ...
+jArray[0] = "hello"; // Possible though a string implicit operator.
+jArray[1] = myCustomDataType; // We can't know all types, so this won't work.
+
+string v0 = jArray[0];
+MyCustomDataType v1 = jArray[1]; // We can't know all types, so this won't work.
+```
+
+Proposed (non-dynamic):
+```cs
+JsonArray jArray = ...
+jArray[0] = new JsonValue("hello");
+jArray[1] = new JsonValue(myCustomDataType);
+
+// Get the internal values based on above.
+string v0 = (string)jArray[0].Value;
+MyCustomDataType v1 = (MyCustomDataType)jArray[1].Value;
+
+// If the type is not known, call GetValue() which supports a possible conversion:
+MyCustomDataType v1 = jArray[1].GetValue<MyCustomDataType>();
+```
+
+Proposed (dynamic):
+```cs
+DynamicJsonArray jArray = ...
+jArray[0] = "hello";
+jArray[1] = myCustomDataType;
+
+string v0 = jArray[0];
+MyCustomDataType v1 = jArray[1];
+```
+
+### Specifying the `JsonSerializerOptions` instance
+The `JsonSerializerOptions` instance needs to be specified in order for any run-time added custom converters to be found and to use `PropertyNameCaseInsensitive` for property lookup with `JsonObject`.
+
+It is cumbersome and error-prone to specify the options instance for every new element or property:
+```cs
+    var jObject = new JsonObject(options)
+    {
+        ["MyString"] = new JsonValue("Hello!", options),
+        ["MyBoolean"] = new JsonValue(false, options),
+        ["MyArray"] = new JsonArray(options)
+        {
+            new JsonValue(2, options),
+            new JsonValue(3, options),
+            new JsonValue(42, options)
+        },
+    }
+```
+so it is possible to omit the options instance for non-root members. When a property or element is added, the options instance is set to the parent's instance:
+```cs
+    var jObject = new JsonObject(options) // options still needed at root level
+    {
+        ["MyString"] = new JsonValue("Hello!"),
+        ["MyBoolean"] = new JsonValue(false),
+        ["MyArray"] = new JsonArray()
+        {
+            new JsonValue(2),
+            new JsonValue(3),
+            new JsonValue(42)
+        },
+    }
+```
+
+It is required that all option instances are the same across a given `JsonNode` tree.
+
+### Using `System.Object` for values
+`JsonNode` has indexers that allow `JsonObject` properties and `JsonArray` elements to be specified. This allows a terse and more performant mode:
+```cs
+    var jObject = new JsonObject(options)
+    {
+        ["Child"]["Array"][0]["Message"] = new JsonValue("Hello!");
+        ["Child"]["Array"][1]["Message"] = new JsonValue("Hello!!");
+    }
```

Comment:
Is this type of unchecked chaining representative of real-world code samples? Making it terse by avoiding the need for too much explicit casts is good, but I am not sure if folks actually navigate the DOM in his hard-coded fashion. What happens if the array doesn't have 2 elements?
What would happen if I did something like:
`["Child"]["Array"][25] = 5`

Also, when it comes to JSON DOM creation from scratch, what about this:
```C#
var jObject = new JsonObject
{
    {
        "Child", new JsonArray()
        {
            { "Message", "Hello!" },
            { "Message", "Hello!!" },
        }
    }
};
```

Reply:
It will throw `ArgumentOutOfRangeException` in that case. The same if `dynamic` is used. This is also consistent with Newtonsoft and requests by the Azure team to support a terse syntax.